File: backend/src/api/schemas/onboarding.py
from pydantic import BaseModel, HttpUrl
from typing import Optional
from datetime import datetime
from src.api.models.onboarding import OnboardingStatus, ShiftTiming

class OnboardingBase(BaseModel):
    pass

class CandidateOnboardingUpdate(BaseModel):
    joining_date: Optional[datetime] = None
    cnic_number: Optional[str] = None
    phone_number: Optional[str] = None
    current_address: Optional[str] = None
    emergency_contact: Optional[str] = None
    bank_name: Optional[str] = None
    bank_iban: Optional[str] = None

class HRJoiningDetailsUpdate(BaseModel):
    reporting_time: Optional[str] = None
    office_location: Optional[str] = None
    shift_timing: Optional[ShiftTiming] = None

# Schemas for dealing with document URLs
class CandidateDocumentUpload(BaseModel):
    doc_front_picture_url: Optional[str] = None
    doc_id_card_url: Optional[str] = None
    doc_salary_slip_url: Optional[str] = None
    doc_experience_letter_url: Optional[str] = None
    doc_educational_documents_url: Optional[str] = None
    doc_police_clearance_url: Optional[str] = None
    doc_resume_url: Optional[str] = None
    doc_additional_files_json: Optional[str] = None

class HROnboardingVerify(BaseModel):
    hr_verified: bool
    reject_reason: Optional[str] = None

class ITOnboardingUpdate(BaseModel):
    it_slack_setup: Optional[bool] = None
    it_gmail_setup: Optional[bool] = None
    it_browser_extensions: Optional[bool] = None
    it_gmail_signature: Optional[bool] = None
    it_bordio_access: Optional[bool] = None
    it_office365_access: Optional[bool] = None

class HRInductionUpdate(BaseModel):
    ind_hr_welcome_session: Optional[bool] = None
    ind_hr_handbook_shared: Optional[bool] = None
    ind_hr_policies_explained: Optional[bool] = None

class ITInductionUpdate(BaseModel):
    ind_it_credentials_provided: Optional[bool] = None
    ind_it_security_induction: Optional[bool] = None

class ManagerInductionUpdate(BaseModel):
    ind_manager_buddy_assigned: Optional[bool] = None
    ind_manager_team_intro: Optional[bool] = None

class OnboardingResponse(BaseModel):
    id: int
    application_id: int
    user_id: int
    status: OnboardingStatus
    
    candidate_name: Optional[str] = None
    email: Optional[str] = None
    job_title: Optional[str] = None
    
    joining_date: Optional[datetime] = None
    reporting_time: Optional[str] = None
    office_location: Optional[str] = None
    shift_timing: Optional[ShiftTiming] = None
    
    # Personal info
    cnic_number: Optional[str] = None
    phone_number: Optional[str] = None
    current_address: Optional[str] = None
    emergency_contact: Optional[str] = None
    bank_name: Optional[str] = None
    bank_iban: Optional[str] = None
    
    doc_front_picture_url: Optional[str] = None
    doc_id_card_url: Optional[str] = None
    doc_salary_slip_url: Optional[str] = None
    doc_experience_letter_url: Optional[str] = None
    doc_educational_documents_url: Optional[str] = None
    doc_police_clearance_url: Optional[str] = None
    doc_resume_url: Optional[str] = None
    doc_additional_files_json: Optional[str] = None
    
    hr_verified: bool
    
    it_slack_setup: bool
    it_gmail_setup: bool
    it_browser_extensions: bool
    it_gmail_signature: bool
    it_bordio_access: bool
    it_office365_access: bool
    
    ind_hr_welcome_session: bool
    ind_hr_handbook_shared: bool
    ind_hr_policies_explained: bool
    ind_it_credentials_provided: bool
    ind_it_security_induction: bool
    ind_manager_buddy_assigned: bool
    ind_manager_team_intro: bool
    
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class OnboardingDocumentResponse(BaseModel):
    id: int
    file_name: str
    file_url: str
    file_type: str
    uploaded_at: datetime

    class Config:
        from_attributes = True

class HRDetailedOnboardingResponse(BaseModel):
    candidate_name: str
    email: str
    job_title: str
    status: str
    documents: list[OnboardingDocumentResponse]

File: frontend/src/lib/api/onboarding.ts
import { apiClient, resolveUrl } from './client';

/** Convert a relative /uploads/... URL to a full backend URL for viewing */
export function getDocumentViewUrl(relativeUrl: string | undefined): string | null {
    if (!relativeUrl) return null;
    return resolveUrl(relativeUrl);
}

export interface OnboardingResponse {
    id: number;
    application_id: number;
    user_id: number;
    status: string;
    joining_date?: string;
    reporting_time?: string;
    office_location?: string;
    shift_timing?: string;
    
    // Personal Info
    cnic_number?: string;
    phone_number?: string;
    current_address?: string;
    emergency_contact?: string;
    bank_name?: string;
    bank_iban?: string;
    
    // Virtual fields from backend
    candidate_name?: string;
    email?: string;
    job_title?: string;
    
    doc_front_picture_url?: string;
    doc_id_card_url?: string;
    doc_salary_slip_url?: string;
    doc_experience_letter_url?: string;
    doc_educational_documents_url?: string;
    doc_police_clearance_url?: string;
    doc_resume_url?: string;
    doc_additional_files_json?: string;
    
    hr_verified: boolean;
    
    it_slack_setup: boolean;
    it_gmail_setup: boolean;
    it_browser_extensions: boolean;
    it_gmail_signature: boolean;
    it_bordio_access: boolean;
    it_office365_access: boolean;
    
    ind_hr_welcome_session: boolean;
    ind_hr_handbook_shared: boolean;
    ind_hr_policies_explained: boolean;
    ind_it_credentials_provided: boolean;
    ind_it_security_induction: boolean;
    ind_manager_buddy_assigned: boolean;
    ind_manager_team_intro: boolean;
}

export interface UploadResponse {
    url: string;
    filename: string;
    size: number;
}

export const onboardingApi = {
    getAll: () => apiClient.get<OnboardingResponse[]>('/onboarding'),
    get: (applicationId: number, token?: string | null) => 
        apiClient.get<OnboardingResponse>(`/onboarding/${applicationId}${token ? `?token=${token}` : ''}`),
    
    initiate: (applicationId: number) => apiClient.post<OnboardingResponse>(`/onboarding/${applicationId}`),
    
    updateCandidateInfo: (applicationId: number, data: any, token?: string | null) => 
        apiClient.put<OnboardingResponse>(`/onboarding/${applicationId}/candidate-date${token ? `?token=${token}` : ''}`, data),
        
    hrSetJoiningDetails: (applicationId: number, data: any) => 
        apiClient.put<OnboardingResponse>(`/onboarding/${applicationId}/hr-joining-details`, data),
        
    updateCandidateDocs: (applicationId: number, data: any, token?: string | null) => 
        apiClient.put<OnboardingResponse>(`/onboarding/${applicationId}/candidate-docs${token ? `?token=${token}` : ''}`, data),

    /** 
     * NEW: Robust multi-file upload for onboarding 
     * Maps to: POST /api/v1/onboarding/{application_id}/upload-documents
     */
    uploadDocuments: async (applicationId: number, files: Record<string, File | null>, token?: string | null): Promise<OnboardingResponse> => {
        const formData = new FormData();
        Object.entries(files).forEach(([key, file]) => {
            if (file) formData.append(key, file);
        });
        return apiClient.post<OnboardingResponse>(`/onboarding/${applicationId}/upload-documents${token ? `?token=${token}` : ''}`, formData);
    },
        
    hrVerify: (applicationId: number, data: { hr_verified: boolean }) => 
        apiClient.put<OnboardingResponse>(`/onboarding/${applicationId}/hr-verify`, data),
        
    itSetupUpdate: (applicationId: number, data: any) => 
        apiClient.put<OnboardingResponse>(`/onboarding/${applicationId}/it-setup`, data),
        
    hrInductionUpdate: (applicationId: number, data: any) => 
        apiClient.put<OnboardingResponse>(`/onboarding/${applicationId}/induction/hr`, data),

    itInductionUpdate: (applicationId: number, data: any) => 
        apiClient.put<OnboardingResponse>(`/onboarding/${applicationId}/induction/it`, data),

    managerInductionUpdate: (applicationId: number, data: any) => 
        apiClient.put<OnboardingResponse>(`/onboarding/${applicationId}/induction/manager`, data),
        
    sendWelcomeEmail: (applicationId: number) => 
        apiClient.post<{ message: string }>(`/onboarding/${applicationId}/send-welcome-email`),

    /** 
     * HR-specific detailed view of onboarding 
     */
    getHrDetails: (applicationId: number) => 
        apiClient.get<any>(`/onboarding/hr/${applicationId}`),

    /** 
     * Upload a document file for a specific application and type
     * Maps to: POST /api/v1/onboarding/{application_id}/upload-documents
     */
    uploadDocument: async (applicationId: number, type: string, file: File, token?: string | null): Promise<OnboardingResponse> => {
        const formData = new FormData();
        formData.append(type, file);
        return apiClient.post<OnboardingResponse>(`/onboarding/${applicationId}/upload-documents${token ? `?token=${token}` : ''}`, formData);
    },
    
    complete: (applicationId: number, token?: string | null) => 
        apiClient.post<OnboardingResponse>(`/onboarding/${applicationId}/complete${token ? `?token=${token}` : ''}`),
};

